src/filedef.cpp

```diff
@@ -1426,7 +1426,7 @@ bool FileDefImpl::generateSourceFile() const
   bool verbatimHeaders = Config_getBool(VERBATIM_HEADERS);
   return !isReference() &&
          (sourceBrowser ||
-           (verbatimHeaders && guessSection(name())==Entry::HEADER_SEC)
+           (verbatimHeaders && (guessSection(name())==Entry::HEADER_SEC || !m_includedByMap.empty()))
```

Comment:
This will work, it looks like... but it'll mean that every file, of any type, that's `#include`d by another file is treated as a header.

That may seem correct, or even tautological, but I'm not sure it entirely is. The problem with `#include` is that it's just a preprocessor directive that pulls in the contents of other files. It doesn't actually say anything about the _nature_ of that file.

There are some coding styles where `.cpp` files get `#include`d into others, for various reasons. (Qt's `moc` sets up generated files that way, as one example.) Even if it's used in a `#include`, I'm not sure it makes sense to treat a `.cpp` file as a header.

It might be better to give `guessSection()` the ability to _optionally_ return `Entry::HEADER_SEC` for filenames with no extension, or for extensionless names found only in certain designated directories.

It's a more complex fix, admittedly, but it might have fewer potentially-unwanted side effects.

Reply:
I think that the problem can be present but that the effects are normally limited by means of the setting of `INCLUDE_FILE_PATTERNS`

Reply:
@albert-github That's the thing, this fix bypasses `FILE_PATTERNS` _and_ `INCLUDE_FILE_PATTERNS`. (necessarily, since `guessSection()` ignores both of those configurations as well. Those configs determine which files Doxygen processes in various ways. _This_ affects whether a file is included as a source file when `VERBATIM_HEADERS` is _on_ but `SOURCE_BROWSER` is off. (In those cases, only headers should have browsable sources.)

`guessSection()` returns `Entry::SOURCE_SEC` or `Entry::HEADER_SEC` (or `0`, if there's no extension or an unknown extension) based _solely_ on a hardcoded set of extension mappings, so it will ignore all of the configuration.

Reply:
I did a quick test which the extended example by adding 2 files (contents analogous to bb) and the lines (in main.,cpp):
```
#include "abs.cba"
#include <xyz.zxy>
```
and the result is:

![image](https://user-images.githubusercontent.com/5223533/219109901-b5360876-2a7d-45c0-a8ff-cf2add6e7dd0.png)

(I made the links red for better visibility).
So these "rubbish" names are not linked and I didn't find any further reference either, so I think the problem does not exist or do you have a, small, example showing the problem?

Reply:
@albert-github 
Is that with this patch, without? If it's _without_, the issue isn't the links, it's the inclusion of the source. What you'll find, if you click those links, is that the "aa.h" link leads to the source of the header file, but the "bb" link leads to the documentation for that file, because the source isn't included. (By using `VERBATIM_HEADERS`, it should be. That's the need for this patch, or some other fix.)

If it's _with_ this patch, the "rubbish files" wouldn't be included because those files still aren't being processed by doxygen at all, which is good. But if I were to have a `#include "somefile.cpp"` in a file, where both `somefile.cpp` and the file that includes it are being processed by Doxygen, then this patch would cause `somefile.cpp` to be treated as a header (so its source would be included in the docs, if `VERBATIM_HEADERS` is enabled but not `SOURCE_BROWSER`). Which is the concern I was raising.

Reply:
(Duh, it's your patch, of course you know the reasons for it. Sorry.)

But here's an example of a project that will be affected by this patch:

[doxy_qt.tar.gz](https://github.com/doxygen/doxygen/files/10752358/doxy_qt.tar.gz)

That's a completely-valid Qt project, it will compile and run. But it's "weird", because (for some reason) the user got it in their head to `#include "MainWindow.cpp" from inside `main.cpp`, instead of using it as a separate source file. (Maybe they wanted to write the class inside `main.cpp`, but it's under a different license. Maybe someone told them it will speed up their builds. Who knows?)

Without this patch, `MainWindow.cpp` is treated as a source file; the `#include "MainWindow.cpp"` line in `main.cpp` is linked to the _documentation_ page, but its source isn't included. With this patch, `MainWindow.cpp` becomes a header file, and its source ends up in the docs.

Reply:
Holy carp! And if you're curious how common it is to `#include "something.cpp"`, [this GitHub code search](https://github.com/search?q=include+%5C.cpp+extension%3Acpp+extension%3Acxx+extension%3Acc+extension%3AC&type=Code&ref=advsearch&l=&l=) returns almost _**40 MILLION**_ code results that broadly fit that pattern. 

(There are surely some false positives, because GitHub code search is "fuzzy" — which is just confounding, in a _code_ search — but regardless, it's clearly _much_ more common than I ever imagined.)

Reply:
Ugh,ignore those numbers. The search is _so_ fuzzy, that one matched every file that contains both "include" and "cpp" _ANYWHERE_ in the text. (And quite possibly, files that contain only "include" _or_ "cpp" as well.)

[This one](https://github.com/search?l=&p=3&q=%22include%2A.cpp%22+extension%3Acpp+extension%3Acxx+extension%3Acc+extension%3AC&ref=advsearch&type=Code) is more realistic, and returns 187,000 matches, which seems much more plausible. And still, plenty common. (Just browsing the first few pages of results, it's clear there are a lot of valid matches.)